xchange-bitfinex/src/main/java/info/bitrich/xchangestream/bitfinex/BitfinexStreamingRawService.java

```diff
@@ -0,0 +1,303 @@
+            return null;
+        }
+
+        long id = order.get(0).longValue();
+        long groupId = order.get(1).longValue();
+        long cid = order.get(2).longValue();
+        String symbol = order.get(3).textValue();
+        long mtsCreate = order.get(4).longValue();
+        long mtsUpdate = order.get(5).longValue();
+        BigDecimal amount = order.get(6).decimalValue();
+        BigDecimal amountOrig = order.get(7).decimalValue();
+        String type = order.get(8).textValue();
+        String typePrev = order.get(9).textValue();
+        int flags = order.get(12).intValue();
+        String orderStatus = order.get(13).textValue();
+        BigDecimal price = order.get(16).decimalValue();
+        BigDecimal priceAvg = order.get(17).decimalValue();
+        BigDecimal priceTrailing = order.get(18).decimalValue();
+        BigDecimal priceAuxLimit = order.get(19).decimalValue();
+        long placedId = order.get(25).longValue();
+
+        return new BitfinexWebSocketAuthOrder(
+                id, groupId, cid, symbol, mtsCreate, mtsUpdate, amount, amountOrig,
+                type, typePrev, orderStatus, price, priceAvg, priceTrailing,
+                priceAuxLimit, placedId, flags
+        );
+    }
+
+    public void auth() {
+        long nonce = System.currentTimeMillis() * 1000;
+        String payload = "AUTH" + nonce;
+        String signature;
+        try {
+            Mac macEncoder = Mac.getInstance(HMAC_SHA_384);
+            SecretKeySpec secretKeySpec = new SecretKeySpec(apiSecret.getBytes(StandardCharsets.UTF_8), HMAC_SHA_384);
+            macEncoder.init(secretKeySpec);
+            byte[] result = macEncoder.doFinal(payload.getBytes(StandardCharsets.UTF_8));
+            signature = DatatypeConverter.printHexBinary(result);
+        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
+            LOG.error("auth. Sign failed error={}", e.getMessage());
+            return;
+        }
+        BitfinexWebSocketAuth message = new BitfinexWebSocketAuth(
+                apiKey, payload, String.valueOf(nonce), signature.toLowerCase()
+        );
+        sendMessage(message);
+    }
+
+
+    public void setApiKey(String apiKey) {
+        this.apiKey = apiKey;
+    }
+
+    public void setApiSecret(String apiSecret) {
+        this.apiSecret = apiSecret;
+    }
+
+    private void sendMessage(Object message) {
+        try {
+            sendMessage(objectMapper.writeValueAsString(message));
+        } catch (JsonProcessingException e) {
+            LOG.error("Error creating json message: {}", e.getMessage());
+        }
+    }
+
+    public Observable<BitfinexWebSocketAuthOrder> getAuthenticatedOrders() {
+        return subjectOrder.share();
+    }
+
+    public Observable<BitfinexWebSocketAuthPreTrade> getAuthenticatedPreTrades() {
+        return subjectPreTrade.share();
+    }
+
+    public Observable<BitfinexWebSocketAuthTrade> getAuthenticatedTrades() {
```

Comment:
I'd like it if we could do as we do for GDAX, and merge the authenticated trades (as `UserTrade`) with the public trades (as `Trade`s) and return them in the `getTrades()` stream.  However, just implementing the L1 API is valuable on its own, so if you don't fancy doing this, that's fine.  We can raise an issue for it and see if someone else picks it up (I might).